### DISCUSSION TITLE
char * ?

### DISCUSSION THREAD

**Discussion by abellgithub** [YosysHQ/yosys#4289] [General]
Can someone explain why RTLIL::IdString uses char * instead of std::string?

**Comment by whitequark** [comment]
It's deduplicated: each `ID::A` refers to the exact same storage, and as such they can be compared by comparing their pointer (extremely cheap) vs comparing their contents (moderately expensive).

**Reply by abellgithub** [reply]
For this to be of value the pointers have to be identical -- meaning that the strings have to be compile-time constant. When the strings are stored in `global_id_storage_` they're dup'd, so the address of the stored string isn't going to be the same as that being looked up unless I'm missing something.

**Reply by abellgithub** [reply]
(They don't *need* to be compile-time constant, but they have to be the same memory.)

**Comment by QuantamHD** [comment]
General technique is called interning https://en.m.wikipedia.org/wiki/String_interning
